### PR TITLE
joint state controller with resource manager

### PR DESCRIPTION
**Pull request by Karsten1987**
Needs https://github.com/ros-controls/ros2_control/pull/236

The way to run it:

terminal1:
```
ros2 run controller_manager ros2_control_node --ros-args -p update_rate:=10 -p "robot_description:=`cat install/test_robot_hardware/share/test_robot_hardware/test_robot_hardware.urdf`" -p "joint_state_controller.type:=joint_state_controller/JointStateController"
```

terminal2:
```
ros2 control load joint_state_controller
```

### CODE REVIEW

File: joint_state_controller/src/joint_state_controller.cpp [hunk 1]
```diff
@@ -38,24 +40,27 @@ class State;
...
 {}
 
-rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
-JointStateController::on_configure(const rclcpp_lifecycle::State & /*previous_state*/)
+controller_interface::InterfaceConfiguration JointStateController::command_interface_configuration() const
 {
-  if (!init_joint_data()) {
-    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
-  }
+  return controller_interface::InterfaceConfiguration{controller_interface::configuration_type::ALL};
```
Comment (destogl) [inline joint_state_controller/src/joint_state_controller.cpp:53]:
```suggestion
  return controller_interface::InterfaceConfiguration{controller_interface::configuration_type::NONE};
```

File: joint_state_controller/include/joint_state_controller/joint_state_controller.hpp [hunk 2]
```diff
@@ -68,6 +65,7 @@ class JointStateController : public controller_interface::ControllerInterface
 
 protected:
   std::vector<std::string> joint_names_;
+  std::unordered_map<std::string, std::unordered_map<std::string, double>> name_if_value_mapping_;
```
Comment (destogl) [inline joint_state_controller/include/joint_state_controller/joint_state_controller.hpp:68]: Can you a short comment about this map?

File: joint_state_controller/src/joint_state_controller.cpp [hunk 3]
```diff
@@ -38,37 +38,55 @@ class State;
 
 namespace joint_state_controller
 {
-static const auto kPositionIndex = 0ul;
-static const auto kVelocityIndex = 1ul;
-static const auto kEffortIndex = 2ul;
+const auto kUninitializedValue = std::numeric_limits<double>::quiet_NaN();
+const auto kPositionName = "position";
+const auto kVelocityName = "velocity";
+const auto kEffortName = "effort";
```
Comment (destogl) [inline joint_state_controller/src/joint_state_controller.cpp:44]: I propose to use names from:  hardware_interface/include/hardware_interface/types/hardware_interface_type_values.hpp

```suggestion
```

File: joint_state_controller/src/joint_state_controller.cpp [hunk 4]
```diff
@@ -18,16 +18,16 @@
 #include <limits>
 #include <memory>
 #include <string>
+#include <unordered_map>
 #include <vector>
 
-#include "hardware_interface/joint_handle.hpp"
-#include "hardware_interface/robot_hardware.hpp"
 #include "hardware_interface/types/hardware_interface_return_values.hpp"
```
Comment (destogl) [inline joint_state_controller/src/joint_state_controller.cpp:24]: Header for standardized names...

```suggestion
#include "hardware_interface/types/hardware_interface_return_values.hpp"
#include "hardware_interface/types/hardware_interface_type_values.hpp"
```

File: joint_state_controller/src/joint_state_controller.cpp [hunk 5]
```diff
@@ -84,93 +102,101 @@ JointStateController::on_deactivate(const rclcpp_lifecycle::State & /*previous_s
...
-          // leave a nullptr for not supported interfaces
-          joint_interface_handles.emplace_back(nullptr);
-        }
-      }
-      joint_handles_.emplace_back(joint_interface_handles);
+  // filter state interfaces that have at least one of the joint_states fields,
+  // the rest will be ignored for this message
+  for (const auto name_ifv : name_if_value_mapping_) {
+    const auto & interfaces_and_values = name_ifv.second;
+    if (has_any_key(interfaces_and_values, {kPositionName, kVelocityName, kEffortName})) {
```
Comment (destogl) [inline joint_state_controller/src/joint_state_controller.cpp:137]:
```suggestion
    if (has_any_key(interfaces_and_values, {hardware_interface::HW_IF_POSITION,
                                                                     hardware_interface::HW_IF_VELOCITY,
                                                                     hardware_interface::HW_IF_EFFORT})) {
```

Maybe not the best formatting possible...

File: joint_state_controller/src/joint_state_controller.cpp [hunk 6]
```diff
@@ -182,38 +208,32 @@ JointStateController::update()
...
-          joint_interface_handle->get_value();
-        ++interface_index;
-      }
-    }
+  for (auto i = 0ul; i < joint_names_.size(); ++i) {
+    joint_state_msg_.position[i] =
+      get_value(name_if_value_mapping_, joint_names_[i], kPositionName);
+    joint_state_msg_.velocity[i] =
+      get_value(name_if_value_mapping_, joint_names_[i], kVelocityName);
+    joint_state_msg_.effort[i] = get_value(name_if_value_mapping_, joint_names_[i], kEffortName);
```
Comment (destogl) [inline joint_state_controller/src/joint_state_controller.cpp:219]:
```suggestion
      get_value(name_if_value_mapping_, joint_names_[i], hardware_interface::HW_IF_POSITION);
    joint_state_msg_.velocity[i] =
      get_value(name_if_value_mapping_, joint_names_[i], hardware_interface::HW_IF_VELOCITY);
    joint_state_msg_.effort[i] = get_value(name_if_value_mapping_, joint_names_[i], hardware_interface::HW_IF_EFFORT);
```

File: joint_state_controller/test/test_common.hpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TEST_COMMON_HPP_
+#define TEST_COMMON_HPP_
+
+
+namespace joint_state_controller_testing
+{
+constexpr auto urdf =
```
Comment (destogl) [inline joint_state_controller/test/test_common.hpp:21]: I would propose to use descriptions.hpp header for this as proposed in #99 and ros-controls/ros2_control#247.
We should just move the file in `test_robot_hardware` as @Karsten1987 proposed.

File: joint_state_controller/test/test_joint_state_controller.cpp [hunk 8]
```diff
@@ -251,10 +268,25 @@ TEST_F(JointStateControllerTest, DynamicJointStatePublishTest)
   rclcpp::MessageInfo msg_info;
   ASSERT_TRUE(subscription->take(dynamic_joint_state_msg, msg_info));
 
-  // checking positions is enough
   const size_t NUM_JOINTS = 3;
-  ASSERT_EQ(dynamic_joint_state_msg.joint_names.size(), NUM_JOINTS);
-  ASSERT_EQ(dynamic_joint_state_msg.interface_values[0].values[0], 1.1);
-  ASSERT_EQ(dynamic_joint_state_msg.interface_values[1].values[0], 2.1);
-  ASSERT_EQ(dynamic_joint_state_msg.interface_values[2].values[0], 3.1);
+  const auto INTERFACE_NAMES = {"position", "velocity", "effort"};
```
Comment (destogl) [inline joint_state_controller/test/test_joint_state_controller.cpp:272]:
```suggestion
  const auto INTERFACE_NAMES = {hardware_interface::HW_IF_POSITION, 
                                                           hardware_interface::HW_IF_VELOCITY, 
                                                           hardware_interface::HW_IF_EFFORT};
```

File: joint_state_controller/test/test_joint_state_controller.cpp [hunk 9]
```diff
@@ -17,19 +17,26 @@
...
 #include <memory>
 #include <string>
+#include <utility>
 #include <vector>
 
-#include "gtest/gtest.h"
+#include "gmock/gmock.h"
 
+#include "hardware_interface/loaned_state_interface.hpp"
 #include "hardware_interface/types/hardware_interface_return_values.hpp"
```
Comment (destogl) [inline joint_state_controller/test/test_joint_state_controller.cpp:26]:
```suggestion
#include "hardware_interface/types/hardware_interface_return_values.hpp"
#include "hardware_interface/types/hardware_interface_type_values.hpp"
```

**Review by destogl — CHANGES_REQUESTED** [review]
Looks good. Just some changes regarding standard constants and shared test data.

**Comment by v-lopez** [comment]
Looks good to me after @destogl suggestions are addressed.

File: joint_state_controller/include/joint_state_controller/joint_state_controller.hpp (hunk 2 above)
Comment (bmagyar) [reply joint_state_controller/include/joint_state_controller/joint_state_controller.hpp:68]: Added comments, I've also grouped the joint-state-related stuff and the dynamic joint state-reltaed stuff together

File: joint_state_controller/src/joint_state_controller.cpp (hunk 3 above)
Comment (bmagyar) [reply joint_state_controller/src/joint_state_controller.cpp:44]: at the time of first doing this, these values were not available. Replaced with this constants now

File: joint_state_controller/src/joint_state_controller.cpp (hunk 4 above)
Comment (bmagyar) [reply joint_state_controller/src/joint_state_controller.cpp:24]: thanks, added

File: joint_state_controller/src/joint_state_controller.cpp (hunk 5 above)
Comment (bmagyar) [reply joint_state_controller/src/joint_state_controller.cpp:137]: when working around linters and code formatters, there's only so much you can discuss about code style :D 

removing the hardware_interface namespace by a couple of `using`s at the top of the file kept this a one-liner

File: joint_state_controller/src/joint_state_controller.cpp (hunk 6 above)
Comment (bmagyar) [reply joint_state_controller/src/joint_state_controller.cpp:219]: ja

File: joint_state_controller/test/test_common.hpp (hunk 7 above)
Comment (bmagyar) [reply joint_state_controller/test/test_common.hpp:21]: https://github.com/ros-controls/ros2_controllers/issues/113 created follow-up issue

File: joint_state_controller/test/test_joint_state_controller.cpp (hunk 9 above)
Comment (bmagyar) [reply joint_state_controller/test/test_joint_state_controller.cpp:26]: done

File: joint_state_controller/test/test_joint_state_controller.cpp (hunk 8 above)
Comment (bmagyar) [reply joint_state_controller/test/test_joint_state_controller.cpp]: yup

Comment (bmagyar) [reply joint_state_controller/test/test_joint_state_controller.cpp]: there was another one of these, done that too

**Comment by bmagyar** [comment]
ok guys, I've addressed all your comments, please verify/resolve discussions (I'm not going to resolve your discussion points)

**Comment by v-lopez** [comment]
I agree with how you addressed @destogl review, could be merged by me.

**Review by bmagyar — APPROVED** [review]
This wasn't my PR originally so I can still "review it"! After doing some major changes to it, the controller works and testing is mostly appropriate. 

_Patting myself on the shoulder in approval_

File: joint_state_controller/src/joint_state_controller.cpp (hunk 5 above)
Comment (destogl) [reply joint_state_controller/src/joint_state_controller.cpp]: Yeah, I mean my proposal was not the best formating possible.

**Review by destogl — APPROVED** [review]
Looks good!

**Pull request merged**